File: TTAPI2013/TTAPIFunctions.cpp
#include <string>
#include "stdafx.h"
#include "TTAPIFunctions.h"


TTAPIFunctions::TTAPIFunctions(System::String ^ username, System::String ^ password)
{
	m_username = username;
	m_password = password;
}

void TTAPIFunctions::Start()
{
	m_disp = TTAPI::Dispatcher::AttachWorkerDispatcher();
	m_disp->BeginInvoke(gcnew System::Action(this, &TTAPIFunctions::Init));
	m_disp->Run();
}

void TTAPIFunctions::Init()
{
	TTAPI::ApiInitializeHandler ^ handler = gcnew TTAPI::ApiInitializeHandler(this, &TTAPIFunctions::handleLogin);
	TTAPI::TTAPI::CreateUniversalLoginTTAPI(TTAPI::Dispatcher::Current, m_username, m_password, handler);
}

void TTAPIFunctions::handleLogin(TTAPI::TTAPI ^ api, TTAPI::ApiCreationException ^ ex)
{
	if (ex == nullptr)
	{
		m_apiInstance = (TTAPI::UniversalLoginTTAPI ^)api;
		m_apiInstance->AuthenticationStatusUpdate += gcnew System::EventHandler<TradingTechnologies::TTAPI::AuthenticationStatusUpdateEventArgs ^>(this, &TTAPIFunctions::OnAuthenticationStatusUpdate);
		m_apiInstance->Start();
		(*m_apiInstance).AuthenticationStatusUpdate
	}
}

TTAPIFunctions::~TTAPIFunctions()
{
}

void TTAPIFunctions::OnAuthenticationStatusUpdate(System::Object ^sender, TradingTechnologies::TTAPI::AuthenticationStatusUpdateEventArgs ^e)
{
	if (e->Status->IsSuccess)
	{
		// lookup an instrument
		System::Console::WriteLine("TT Login suceeded, looking up instrument");
		m_req = gcnew TTAPI::InstrumentLookupSubscription(m_apiInstance->Session, TTAPI::Dispatcher::Current, TTAPI::ProductKey(TTAPI::MarketKey::Cme, TTAPI::ProductType::Future, "ES"), "Mar16");
		m_req->Update += gcnew System::EventHandler<TradingTechnologies::TTAPI::InstrumentLookupSubscriptionEventArgs ^>(this, &TTAPIFunctions::OnUpdate);
		m_req->Start();
	}
}


void TTAPIFunctions::OnUpdate(System::Object ^sender, TradingTechnologies::TTAPI::InstrumentLookupSubscriptionEventArgs ^e)
{
	if (e->Instrument != nullptr && e->Error == nullptr)
	{
		// Instrument was found
		System::Console::WriteLine("Found: {0}", e->Instrument->Name);

		// Subscribe for Inside Market Data
		m_ps = gcnew TTAPI::PriceSubscription(e->Instrument, TTAPI::Dispatcher::Current);
		m_ps->Settings = gcnew TTAPI::PriceSubscriptionSettings(TTAPI::PriceSubscriptionType::InsideMarket);
		m_ps->FieldsUpdated += gcnew TradingTechnologies::TTAPI::FieldsUpdatedEventHandler(this, &TTAPIFunctions::OnFieldsUpdated);
		m_ps->Start();
	}
	else if (e->IsFinal)
	{
		// Instrument was not found and TT API has given up looking for it
		System::Console::WriteLine("Cannot find instrument: {0}", e->Error->Message);
	}
}


void TTAPIFunctions::OnFieldsUpdated(System::Object ^sender, TradingTechnologies::TTAPI::FieldsUpdatedEventArgs e)
{
	if (e.Error == nullptr)
	{
		if (e.UpdateType == TTAPI::UpdateType::Snapshot)
		{
			// Received a market data snapshot
			System::Console::WriteLine("Market Data Snapshot:");

			for each(TTAPI::FieldId id in e.Fields->GetFieldIds())
			{
				if (id == TTAPI::FieldId::SettlementPrice)
				{
					System::Console::WriteLine("Settlement Price");
				}
				System::Console::WriteLine("    {0} : {1}", id.ToString(), e.Fields[id]->FormattedValue);
			}
		}
		else
		{
			// Only some fields have changed
			System::Console::WriteLine("Market Data Update:");

			for each(TTAPI::FieldId id in e.Fields->GetChangedFieldIds())
			{
				if (id == TTAPI::FieldId::SettlementPrice)
				{
					System::Console::WriteLine("Settlement Price Changed");
				}
				System::Console::WriteLine("    {0} : {1}", id.ToString(), e.Fields[id]->FormattedValue);
			}
			TTAPI::TradingStatus ts = e.Fields->GetSeriesStatusField()->Value;
		}
	}
	else
	{
		if (e.Error->IsRecoverableError == false)
		{
			System::Console::WriteLine("Unrecoverable price subscription error: {0}", e.Error->Message);
			//Dispose();
		}
	}
}

File: TimeAndSalesOrganizerC#/ConsoleApplication3/TTAPIFunctions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using TimeAndSalesOrganizer;

namespace TTAPI_Init_Worker_UL
{
    using TradingTechnologies.TTAPI;
    // Work notes
    // Working currently on making seperate "Record timings"
    // 1. Somehow set standard format for time record
    // 2. Create parser
    // 3. Create tokenizer for the sample application input
    // 4. Create application

    /// <summary>
    /// Main TT API class
    /// </summary>
    class TTAPIFunctions : IDisposable
    {
        /// <summary>
        /// Declare the TTAPI objects
        /// </summary>
        private UniversalLoginTTAPI m_apiInstance = null;
        private WorkerDispatcher m_disp = null;
        private bool m_disposed = false;
        private object m_lock = new object();
        private System.IO.StreamWriter file;
        private string startTime;
        private Boolean firstTime = true;


        /// <summary>
        /// Private default constructor
        /// </summary>
        public TTAPIFunctions()
        {
        }

        /// <summary>
        /// Create and start the Dispatcher
        /// </summary>
        public void Start()
        {
            // Attach a WorkerDispatcher to the current thread
            m_disp = Dispatcher.AttachWorkerDispatcher();
            m_disp.BeginInvoke(new Action(Init));
            m_disp.Run();
        }

        /// <summary>
        /// Initialize TT API
        /// </summary>
        public void Init()
        {
            // Use "Universal Login" Login Mode
            ApiInitializeHandler h = new ApiInitializeHandler(ttApiInitComplete);
            TTAPI.CreateUniversalLoginTTAPI(Dispatcher.Current, AppProperties.m_username, AppProperties.m_password, h);
        }

        /// <summary>
        /// Event notification for status of TT API initialization
        /// </summary>
        public void ttApiInitComplete(TTAPI api, ApiCreationException ex)
        {
            if (ex == null)
            {
                // Authenticate your credentials
                m_apiInstance = (UniversalLoginTTAPI)api;
                m_apiInstance.AuthenticationStatusUpdate += new EventHandler<AuthenticationStatusUpdateEventArgs>(apiInstance_AuthenticationStatusUpdate);
                m_apiInstance.Start();
            }
            else
            {
                Console.WriteLine("TT API Initialization Failed: {0}", ex.Message);
                Dispose();
            }
        }

        /// <summary>
        /// Event notification for status of authentication
        /// </summary>
        public void apiInstance_AuthenticationStatusUpdate(object sender, AuthenticationStatusUpdateEventArgs e)
        {
            if (e.Status.IsSuccess)
            {
                // Add code here to begin working with the TT API
                InstrumentLookupSubscription instSub = new InstrumentLookupSubscription(m_apiInstance.Session, Dispatcher.Current,
                  AppProperties.prod1Key, AppProperties.contract1Key);
                instSub.Update += new EventHandler<InstrumentLookupSubscriptionEventArgs>(instrumentUpdate);
                instSub.Start();
            }
            else
            {
                Console.WriteLine("TT Login failed: {0}", e.Status.StatusMessage);
                Dispose();
            }
        }

        void instrumentUpdate(object sender, InstrumentLookupSubscriptionEventArgs e)
        {
            if (e.Instrument != null && e.Error == null)
            {
                
                PriceSubscription ps = new PriceSubscription(e.Instrument, Dispatcher.Current);
                ps.Settings = new PriceSubscriptionSettings(PriceSubscriptionType.InsideMarket);
                ps.FieldsUpdated += new FieldsUpdatedEventHandler(ps_FieldsUpdated);
                ps.Start();

                TimeAndSalesSubscription tsSub = new TimeAndSalesSubscription(e.Instrument, Dispatcher.Current);
                tsSub.Update += new EventHandler<TimeAndSalesEventArgs>(timeAndSalesUpdate);
                tsSub.Start();
            }
            else if (e.IsFinal)
            {
                // Instrument was not found and TT API has given up looking for it
            }
            else
            {
                // Instrument was not found yet - but still searching
            }
        }

        void timeAndSalesUpdate(object sender, TimeAndSalesEventArgs e)
        {
            string write = "";
            foreach (TimeAndSalesData tsData in e.Data)
            {
                DateTime ltt = tsData.TimeStamp;
                Price ltp = tsData.TradePrice;
                Quantity ltq = tsData.TradeQuantity;

                write += (ltt.ToFileTimeUtc() - this.startTime) + "," + ltp.ToString() + "," + ltq.ToString() + "\n";

                if(firstTime)
                {
                    this.startTime = ltt.ToLongDateString();
                    firstTime = false;
                }
     
            }
            using (file = new System.IO.StreamWriter(AppProperties.fileLoc, true))
            {
                file.Write(write);
                
                file.WriteLine("Init " + System.DateTime.Now);
                file.WriteLine("Init " + this.startTime);
            }
            file.Close();
        }

        void ps_FieldsUpdated(object sender, FieldsUpdatedEventArgs e)
        {
            if (e.Error == null)
            {
                Console.WriteLine("Instrment: {0}", e.Fields.Instrument.Name);
                Price lp = (Price)e.Fields[FieldId.LastTradedPrice].Value;
                Console.WriteLine("Price: {0}\nOffsetPrice: {1}", lp.ToString(), lp.GetTickPrice(2).ToString());
            }
        }

        /// <summary>
        /// Shuts down the TT API
        /// </summary>
        public void Dispose()
        {
            lock (m_lock)
            {
                if (!m_disposed)
                {
                    // Begin shutdown the TT API
                    TTAPI.ShutdownCompleted += new EventHandler(TTAPI_ShutdownCompleted);
                    TTAPI.Shutdown();

                    m_disposed = true;
                }
            }
        }

        /// <summary>
        /// Event notification for completion of TT API shutdown
        /// </summary>
        public void TTAPI_ShutdownCompleted(object sender, EventArgs e)
        {
            // Shutdown the Dispatcher
            if (m_disp != null)
            {
                m_disp.BeginInvokeShutdown();
                m_disp = null;
            }

            // Dispose of any other objects / resources
        }
    }
}